src/operator/fusion/fused_op-inl.h

```diff
@@ -0,0 +1,906 @@
+  return 1.f/(1 + expf(-val));
+}
+
+template <typename DType>
+inline DType backward_sigmoid(const DType out, const DType grad) {
+  return grad * out * (1 - out);
+}
+
+template <typename DType>
+inline DType softrelu(const DType val) {
+  return logf(1 + expf(val));
+}
+
+template <typename DType>
+inline DType backward_softrelu(const DType val, const DType grad) {
+  return grad * sigmoid(val);
+}
+
+template <typename DType>
+inline DType softsign(const DType val) {
+  return val / (1 + fabsf(val));
+}
+
+template <typename DType>
+inline DType backward_softsign(const DType val, const DType grad) {
+  const DType ap1 = 1 + fabsf(val);
+  return grad / (ap1 * ap1);
+}
+
+// exp and log
+
+template <typename DType>
+inline DType exp(const DType val) {
+  return expf(val);
+}
+
+template <typename DType>
+inline DType backward_exp(const DType val, const DType grad) {
+  return grad * expf(val);
+}
+
+template <typename DType>
+inline DType expm1(const DType val) {
+  return expm1f(val);
+}
+
+template <typename DType>
+inline DType backward_expm1(const DType val, const DType grad) {
+  return grad * expf(val);
+}
+
+template <typename DType>
+inline DType log(const DType val) {
+  return logf(val);
+}
+
+template <typename DType>
+inline DType backward_log(const DType val, const DType grad) {
+  return grad / val;
+}
+
+template <typename DType>
+inline DType log10(const DType val) {
+  return log10f(val);
+}
+
+template <typename DType>
+inline DType backward_log10(const DType val, const DType grad) {
+  return grad / (val * logf(10));
+}
+
+template <typename DType>
+inline DType log2(const DType val) {
+  return log2f(val);
+}
+
+template <typename DType>
+inline DType backward_log2(const DType val, const DType grad) {
+  return grad / (val * logf(2));
+}
+
+template <typename DType>
+inline DType log1p(const DType val) {
+  return log1pf(val);
+}
+
+template <typename DType>
+inline DType backward_log1p(const DType val, const DType grad) {
+  return grad / (1 + val);
+}
+
+// trigonometric
+
+constexpr double pi = 3.14159265358979323846;
+
+template <typename DType>
+inline DType degrees(const DType val) {
+  return (val / pi) * 180;
+}
+
+template <typename DType>
+inline DType radians(const DType val) {
+  return (val / 180.0) * pi;
+}
+
+template <typename DType>
+inline DType sin(const DType val) {
+  return sinf(val);
+}
+
+template <typename DType>
+inline DType backward_sin(const DType val, const DType grad) {
+  return grad * cosf(val);
+}
+
+template <typename DType>
+inline DType cos(const DType val) {
+  return cosf(val);
+}
+
+template <typename DType>
+inline DType backward_cos(const DType val, const DType grad) {
+  return -grad * sinf(val);
+}
+
+template <typename DType>
+inline DType tan(const DType val) {
+  return tanf(val);
+}
+
+// Uses output from tan
+template <typename DType>
+inline DType backward_tan(const DType out, const DType grad) {
+  return grad * (out * out + 1);
+}
+
+template <typename DType>
```

Comment:
stupid question, but why can't we use mshadow ops for this? (mshadow_op.h)   why reimplement?

Reply:
Main reason is that it is not really possible to include mshadow_op.h in runtime compilation - we would need to ship A LOT of headers with binary package (including entire mshadow) and make a mechanism for NVRTC to be able to find them. mshadow_op.h also relies on cuda_fp16.h, which would mean that we need to be able to find at runtime CUDA include directory.